Feature/nex 136/configure spanner

julien-sebire opened the pull request:
This requires PR https://github.com/oat-sa/lib-generis-search/pull/21 and release of lib-generis-search.

This adds NewSql schema support.

jbout committed:
Use DBAL now expression

jbout committed:
Move to a uuid statements primary key

julien-sebire committed:
Fixed statements table creation.

julien-sebire committed:
Fixed schema and typos...

bugalot committed:
Changes to make it work a little bit more.

bugalot committed:
Merge branch 'feature/autoincrement-less-statements' of https://github.com/oat-sa/generis into feature/autoincrement-less-statements

julien-sebire committed:
Fixed wrong call to uuid generator.

julien-sebire committed:
Changed hard-coded value for writable model id with a key generated with the same method than the other ones.

julien-sebire committed:
Some codestyle.

julien-sebire committed:
Got rid of gcp Spanner reference in comments.

julien-sebire committed:
PR fixes.

julien-sebire committed:
Fixed insert return value testing.

julien-sebire committed:
Use the configured writable model id instead of hardcoded value.

julien-sebire committed:
Ensure that the searched model uri has a trailing #.

bugalot committed:
Merge branch 'develop' into feature/autoincrement-less-statements

julien-sebire committed:
Upgraded dbal to version 2.9 and required doctrine/annotations.

julien-sebire committed:
Added spanner dbal driver.

julien-sebire committed:
Added minimum stability dev to get lib-dbal-spanner:dev-develop.

julien-sebire committed:
Required branch feature/NEX-84/spanner-query-refactoring of lib-dbal-spanner

julien-sebire committed:
Adding minimum stability:dev.

julien-sebire committed:
Merge remote-tracking branch 'origin/feature/NEX-121/upgrade-php71' into feature/NEX-136/configure-spanner

# Conflicts:
#	manifest.php
#	scripts/update/Updater.php

julien-sebire committed:
Merge remote-tracking branch 'origin/fix/now-expression' into feature/NEX-136/configure-spanner

julien-sebire committed:
Merge remote-tracking branch 'origin/feature/autoincrement-less-statements' into feature/NEX-136/configure-spanner

julien-sebire committed:
requiring lib-search added quoting of modelId.

julien-sebire committed:
Merge branch 'feature/NEX-136/configure-spanner' of github.com:oat-sa/generis into feature/NEX-136/configure-spanner

jbout reviewed (changes requested):
Should not be merged as it is.
Some of the changes would be mergable already, but it would require a split into multiple PRs.

ferenckiss89 commented:
> Should not be merged as it is.
> Some of the changes would be mergable already, but it would require a split into multiple PRs.

Please do not consider the previous comments now.

ekkinox commented on composer.json:
```diff
@@ -53,16 +53,24 @@
     "league/flysystem": "~1.0",
     "oat-sa/oatbox-extension-installer": "~1.1||dev-master",
-    "php": ">=5.6.6",
+    "php": "^7.1",
```
good to see :)

julien-sebire replied:
YES!!!

ekkinox commented on core/kernel/api/class.ModelFactory.php:
```diff
@@ -16,112 +17,146 @@
+     *
+     * @return string
+     */
+    public function getUniquePrimaryKey()
```
this stuff is in a lot of places, maybe consider a trait, of if you can, use DI for a tiny generator (will ease your life once you'll start to test)

julien-sebire replied:
Right, moved to oat\generis\Helper\UuidPrimaryKeyTrait

ekkinox commented on core/kernel/persistence/smoothsql/install/SmoothRdsModel.php:
```diff
@@ -17,43 +17,49 @@
-        $table->addColumn('modeluri', "string", array("length" => 255,"default" => null));
+        // Models table.
+        $table = $schema->createTable('models');
+        $table->addColumn('modelid', 'string', ['length' => 25, 'notnull' => true]);
```
again, the 23 vs 25 topic

julien-sebire replied:
Fixed.

julien-sebire committed:
Moved Uuid primary key generation to oat\generis\Helper\UuidPrimaryKeyTrait.

julien-sebire committed:
Merge remote-tracking branch 'origin/feature/NEX-136/configure-spanner' into feature/NEX-136/configure-spanner

ferenckiss89 commented on common/ext/class.Namespace.php:
```diff
@@ -1,169 +1,117 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2008-2010 (original work) Deutsche Institut für Internationale Pädagogische Forschung (under the project TAO-TRANSFER);
+ *               2009-2012 (update and modification) Public Research Centre Henri Tudor (under the project TAO-SUSTAIN & TAO-DEV);
```
Why we are not updating the dates here?

peetya commented on core/kernel/persistence/smoothsql/install/SmoothRdsModel.php:
```diff
@@ -17,43 +17,49 @@
+        $table = $schema->createTable('statements');
+        $table->addColumn('id', 'string', ['length' => 25, 'notnull' => true]);
+
+        $table->addColumn('modelid', 'string', ['length' => 25, 'notnull' => true]);
```
Maybe the length could be changed here as well, as Jonathan recommended?

julien-sebire replied:
Right, I missed this one. Done.

peetya commented on scripts/update/Updater.php:
```diff
@@ -430,6 +430,6 @@ public function update($initialVersion)
         }
 
-        $this->skip('11.3.2', '11.4.2');
+        $this->skip('11.3.2', '12.0.0');
```
I think the Updater should care about updating the table schema. What do you think?

julien-sebire replied:
Yes, but this will need a bigger work, since we'd have to deal with existing tao instances. We had a small talk with Joel about it last night and this won't be merged as soon...

julien-sebire committed:
Changed Uuid storing size to 23 bytes.

julien-sebire committed:
Added enclosing quotes for model id.

julien-sebire committed:
Merge remote-tracking branch 'origin/feature/NEX-136/configure-spanner' into feature/NEX-136/configure-spanner

julien-sebire committed:
Merged develop.

julien-sebire committed:
Opened statement type hinting to SpannerStatement also.

julien-sebire committed:
Casted object property to string (Spanner is strict on field types).

julien-sebire committed:
Added temporary LIKE comparator for Spanner.
It is case sensitive so not equivalent to the ones for other platforms.
TODO: For case-insensitive version, we need to use LOWER(expression) LIKE LOWER(expression) or the same with UPPER.

julien-sebire committed:
Merge remote-tracking branch 'origin/develop' into feature/NEX-136/configure-spanner

jbout committed:
Merge branch 'develop' into feature/NEX-136/configure-spanner

julien-sebire committed:
Merge branch 'develop' into feature/NEX-136/configure-spanner

julien-sebire committed:
Removed Spanner dependency for AWS benchmark.

julien-sebire committed:
Removed Spanner class in type hint for aws benchmark.

julien-sebire committed:
Casted boolean values to integer in prepared statement for postgres dbal driver bug (see https://github.com/doctrine/dbal/issues/1847).

julien-sebire committed:
Set bin2hex as default uri provider.

julien-sebire committed:
Added types to all queries.

julien-sebire committed:
Removed wrong fix for postgres boolean issue.

julien-sebire committed:
Merged develop back into NEX-136

julien-sebire committed:
Added abstract changes to enable schema changes.

julien-sebire committed:
Merge branch 'feature/NEX-136/configure-spanner' of https://github.com/oat-sa/generis into feature/NEX-136/configure-spanner

julien-sebire committed:
Merge branch 'develop' into feature/NEX-136/configure-spanner

julien-sebire committed:
Removed forgotten TODOs.

julien-sebire committed:
Merge branch 'feature/NEX-136/configure-spanner' of https://github.com/oat-sa/generis into feature/NEX-136/configure-spanner

jbout commented on composer.json:
```diff
@@ -53,6 +53,12 @@
   "extra": {
     "tao-extension-name": "generis"
   },
+  "repositories": [
```
repositories added in non root composer.json will be ignored an lead to confussion

jbout commented on composer.json:
```diff
@@ -96,5 +102,6 @@
     }
-  }
+  },
+  "minimum-stability" : "dev"
```
minimum-stability is ignored in non root composer.json and can be confussing

julien-sebire committed:
Added an abstraction level for ModelFactory.

julien-sebire committed:
Revert CRLF change.

julien-sebire committed:
Merge branch 'develop' into feature/NEX-136/configure-spanner

julien-sebire committed:
Merge branch 'develop' into feature/NEX-136/configure-spanner

julien-sebire committed:
Updated composer.json to require changes in lib-generis-search.

julien-sebire committed:
Removed unrelated commits.

julien-sebire committed:
Reverted CRLF changes in Namespace.

mike-ionut-mihai-sandu-tao commented on core/kernel/api/class.ModelFactory.php:
```diff
@@ -16,112 +17,181 @@
-            //TODO bad way, need to find better
-            $modelId = $this->getModelId($namespace);
+        if ($modelId === false) {
+            common_Logger::d('modelId not found, need to add namespace ' . $namespace);
```
can we use the LoggerAwareTrait and the function from inside that instead of this singleton old logger

mike-ionut-mihai-sandu-tao commented on core/kernel/api/class.ModelFactory.php:
```diff
@@ -16,112 +17,181 @@
-                )
-            );
+
+        if (intval($result->fetchColumn()) > 0) {
```
casting to int will optimize your condition

mike-ionut-mihai-sandu-tao commented on core/kernel/api/class.ModelFactory.php:
```diff
@@ -16,112 +17,181 @@
+    public function getIteratorQuery($modelIds)
+    {
+        return 'SELECT * FROM statements '
+            . (is_null($modelIds) ? '' : 'WHERE ' . $this->buildModelSqlCondition($modelIds) . ' ')
```
i don't think you need the `is_null`, also i think a `sprintf` will be more readable

mike-ionut-mihai-sandu-tao commented on core/kernel/impl/class.ApiModelOO.php:
```diff
@@ -47,22 +46,43 @@ class core_kernel_impl_ApiModelOO
-    private static $instance = null;
+    public static function singleton()
+    {
+        $returnValue = null;
```
normally 
```
    if (self::$instance == null)
    {
      self::$instance = new Singleton();
    }
 
    return self::$instance;
```
but your implementation is different, can you please comment on the value of this form

mike-ionut-mihai-sandu-tao commented on helpers/UuidPrimaryKeyTrait.php:
```diff
@@ -0,0 +1,37 @@
+     */
+    public function getUniquePrimaryKey()
+    {
+        return strrev(uniqid('', true));
```
we will modify this with uuid, no ?

julien-sebire replied:
Sure we do.

julien-sebire committed:
PR fixes.

julien-sebire committed:
Moved smooth tests to integration.

julien-sebire committed:
FQCN fixes + code style.

julien-sebire committed:
Independent tests.

julien-sebire committed:
Removed getUniquePrimaryKey from persistence, provided by helper.

julien-sebire committed:
Merge branch 'feature/NEX-136/configure-spanner' of https://github.com/oat-sa/generis into feature/NEX-136/configure-spanner

julien-sebire reviewed (commented):
Updated PR.

julien-sebire committed:
Model ids as md5 of namespaces.

jbout commented on common/ext/class.GenerisInstaller.php:
```diff
@@ -48,16 +40,21 @@ public function install()
-            \core_kernel_persistence_smoothsql_SmoothModel::OPTION_READABLE_MODELS => array('1'),
-            \core_kernel_persistence_smoothsql_SmoothModel::OPTION_WRITEABLE_MODELS => array('1'),
-            \core_kernel_persistence_smoothsql_SmoothModel::OPTION_NEW_TRIPLE_MODEL => '1',
+            \core_kernel_persistence_smoothsql_SmoothModel::OPTION_READABLE_MODELS => [$writableModelId],
```
already part of https://github.com/oat-sa/generis/pull/686/files#diff-cb45d9e73f3ab29ea11eadbdc095c55fR59

julien-sebire committed:
Removed GenerisInstaller, already in other PR.

julien-sebire committed:
Replaced dbWrapper with Persistence.
